File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/utils/dialogs/ProgressDialogFragment.java
package com.github.anrimian.musicplayer.ui.utils.dialogs;

import android.app.AlertDialog;
import android.app.Dialog;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.fragment.app.DialogFragment;

import com.github.anrimian.musicplayer.Constants;
import com.github.anrimian.musicplayer.R;

import static com.github.anrimian.musicplayer.Constants.Tags.MESSAGE_ARG;
import static com.github.anrimian.musicplayer.Constants.Tags.MESSAGE_RES_ARG;

public class ProgressDialogFragment extends DialogFragment {

    public static ProgressDialogFragment newInstance(String message) {
        Bundle args = new Bundle();
        args.putString(MESSAGE_ARG, message);
        ProgressDialogFragment fragment = new ProgressDialogFragment();
        fragment.setArguments(args);
        return fragment;
    }

    public static ProgressDialogFragment newInstance(@StringRes int messageResId) {
        Bundle args = new Bundle();
        args.putInt(MESSAGE_RES_ARG, messageResId);
        ProgressDialogFragment fragment = new ProgressDialogFragment();
        fragment.setArguments(args);
        return fragment;
    }

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setCancelable(false);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        View view = View.inflate(getActivity(), R.layout.dialog_progress, null);

        TextView tvProgress = view.findViewById(R.id.tv_progress);
        tvProgress.setText(getMessage());

        return new AlertDialog.Builder(getActivity())
                .setView(view)
                .setCancelable(false)
                .create();
    }

    private String getMessage() {
        Bundle args = getArguments();
        assert args != null;
        int resId = args.getInt(MESSAGE_RES_ARG);
        if (resId != 0) {
            return getString(resId);
        }
        return args.getString(MESSAGE_ARG);
    }
}

File: app/src/main/java/com/github/anrimian/musicplayer/ui/utils/dialogs/ProgressDialogFragment.kt
package com.github.anrimian.musicplayer.ui.utils.dialogs

import android.app.AlertDialog
import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.fragment.app.DialogFragment
import com.github.anrimian.musicplayer.Constants
import com.github.anrimian.musicplayer.R
import com.github.anrimian.musicplayer.ui.common.format.indeterminate
import com.github.anrimian.musicplayer.ui.common.format.setExtProgress
import com.google.android.material.progressindicator.CircularProgressIndicator

fun newProgressDialogFragment(message: String?) = ProgressDialogFragment().apply {
    arguments = Bundle().apply {
        putString(Constants.Tags.MESSAGE_ARG, message)
    }
}

fun newProgressDialogFragment(@StringRes messageResId: Int) = ProgressDialogFragment().apply {
    arguments = Bundle().apply {
        putInt(Constants.Tags.MESSAGE_RES_ARG, messageResId)
    }
}

class ProgressDialogFragment : DialogFragment() {

    private lateinit var progressBar: CircularProgressIndicator
    private lateinit var tvProgress: TextView

    private var cancelListener: (() -> Unit)? = null

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val view = View.inflate(activity, R.layout.dialog_progress, null)

        progressBar = view.findViewById(R.id.progressBar)
        tvProgress = view.findViewById(R.id.tvProgress)

        tvProgress.text = getMessage()

        isCancelable = cancelListener != null

        return AlertDialog.Builder(activity)
            .setView(view)
            .apply {
                if (cancelListener != null) {
                    setNegativeButton(R.string.cancel) { dialog, _ -> dialog.cancel() }
                }
            }
            .create()
    }

    override fun onCancel(dialog: DialogInterface) {
        super.onCancel(dialog)
        cancelListener?.invoke()
    }

    fun setMessage(message: String?) {
        tvProgress.text = message
        requireArguments().putString(Constants.Tags.MESSAGE_ARG, message)
    }

    fun setIndeterminate(indeterminate: Boolean) {
        progressBar.indeterminate(indeterminate)
    }

    fun setProgress(progress: Int) {
        progressBar.setExtProgress(progress)
    }

    fun setCancellationListener(listener: () -> Unit) {
        this.cancelListener = listener
    }

    private fun getMessage(): String? {
        val args = requireArguments()
        val message = args.getString(Constants.Tags.MESSAGE_ARG)
        return if (message == null) {
            val resId = args.getInt(Constants.Tags.MESSAGE_RES_ARG)
            if (resId == 0) {
                null
            } else {
                getString(args.getInt(Constants.Tags.MESSAGE_RES_ARG))
            }
        } else {
            message
        }
    }
}